source/core/slang-performance-profiler.cpp

```diff
@@ -39,11 +40,104 @@ namespace Slang
+                // }
+                // else if (!strcmp(func.key, "propagateConstExpr"))
+                // {
+                //     m_propagateConstExpr += ms;
+                // }
+                // else if (!strcmp(func.key, "performMandatoryEarlyInlining"))
+                // {
+                //     m_earlyInlneTime += ms;
+                // }
+                // else if (!strcmp(func.key, "eliminateDeadCode"))
+                // {
+                //     m_eliminateDeadCode += ms;
+                // }
+                else if (!strcmp(func.key, "linkAndOptimizeIR"))
+                {
+                    m_linkIRTime += ms;
+                }
+                else if (!strcmp(func.key, "linkIR"))
+                {
+                    m_optimizeIRTime += ms;
+                }
+            }
+
+            out << "Semantics Check Time: " << SerialPerformaceProfilerImpl::m_semanticsCheckTime.load() << "ms\n";
+            // out << "IR Generation Time: " << SerialPerformaceProfilerImpl::m_IRGenTime.load() << "ms\n";
+            // out << "    synthesizeBitFieldAccessors time: " << SerialPerformaceProfilerImpl::m_synthesizeBitFieldAccessors.load() << "ms\n";
+            // out << "    constructSSA time: " << SerialPerformaceProfilerImpl::m_constructSSA.load() << "ms\n";
+            // out << "    simplifyCFG time: " << SerialPerformaceProfilerImpl::m_simplifyCFG.load() << "ms\n";
+            // out << "    peepholeOptimize time: " << SerialPerformaceProfilerImpl::m_peepholeOptimize.load() << "ms\n";
+            // out << "    propagateConstExpr time: " << SerialPerformaceProfilerImpl::m_propagateConstExpr.load() << "ms\n";
+            // out << "    eliminateDeadCode time: " << SerialPerformaceProfilerImpl::m_eliminateDeadCode.load() << "ms\n";
+            // out << "    Early Inline Time: " << SerialPerformaceProfilerImpl::m_earlyInlneTime.load() << "ms\n";
+            out << "Link IR Time: " << SerialPerformaceProfilerImpl::m_linkIRTime.load() << "ms\n";
+            out << "Optimize IR Time: " << SerialPerformaceProfilerImpl::m_optimizeIRTime.load() << "ms\n";
+       }
+    private:
+        // Those are supposed to accumulate the time spent in each thread
+        std::atomic<long> m_semanticsCheckTime;
+        std::atomic<long> m_IRGenTime;
+        std::atomic<long> m_earlyInlneTime;
+        std::atomic<long> m_linkIRTime;
+        std::atomic<long> m_optimizeIRTime;
+        std::atomic<long> m_synthesizeBitFieldAccessors;
+        std::atomic<long> m_constructSSA;
+        std::atomic<long> m_simplifyCFG;
+        std::atomic<long> m_peepholeOptimize;
+        std::atomic<long> m_propagateConstExpr;
+        std::atomic<long> m_eliminateDeadCode;
+    };
+
+    // SerialPerformaceProfilerImpl profiler is intent to be shared by all threads, it's thread safe object
+    // because the only member function is re-entrant function, all the member variables are atomic.
```

Comment:
Can we simply require the application to take a mutex lock before entering the critical section/function?
I am not sure using multiple automic objects is better than the previous code that was taking a single mutex_lock.
If the application will take a mutex_lock before entering the function, we are using atomic objects for no good reason and it comes with a certain amount of cost.

Reply:
We can just close this PR now.
This PR is only for debug purpose, so I marked it as Draft. But still thanks for the review.